### DISCUSSION TITLE
Config injection

### DISCUSSION THREAD

**Discussion by MelchiorKokernoot** [laravel/framework#46227] [Ideas]
I have been working on a poc where I am injecting config values through constructor arguments, like so:

```
class Foo implements AutowiresConfigs{
    public function __construct(
        public StringConfig $appName,
    ){}
}
```
This will automatically inject the value of `config('app.name')` into `$appName`. 
I have also created a way to do this with annotations like so:

```
class Foo implements AutowiresConfigs{
    public function __construct(
        #[StringConfig('app.name')]
        public StringConfig $appName,
    ){}
}
```

currently the behaviour is only triggered on classes that extend `AutowiresConfigs` but it is also possible to get here for any class.
Finally I have also found a way to do this with primitive types, so:

```
class Foo implements AutowiresConfigs{
    public function __construct(
        public string$appName,
    ){}
}
```

The package can be found [here](https://github.com/MelchiorKokernoot/laravel-autowire-config) for further inspection, but I am really interested what the general opinion is about this. Should I create a PR so this is supported by the framework?

Kind regards, 

Melchior

**Comment by bram-pkg** [comment]
How does this work for config files with values like `foo-bar.key-one`?

**Reply by henzeb** [reply]
That's where the annotations come in, I guess?

**Reply by MelchiorKokernoot** [reply]
Absolutely! Annotations fully support these keys, and from the package perspective, the callback that should be used to resolve the config key could also be made configurable, with a sensible default of course that is.

**Comment by Radiergummi** [comment]
I would remove the name-based resolution and go forward with the attributes; that's exactly the stuff they were incorporated into the language for, and looks very Laravel-y to me!

**Reply by MelchiorKokernoot** [reply]
Thanks for your reply! I am going to submit my PR as soon as possible and will stick to the attribute way, because it is less 'magicky'.

**Comment by joelharkes** [comment]
Cool! i think this would be a great default Laravel feature. I was looking for it but couldn't find it in the documentation anywhere.

It would make components even more testable.

**Reply by henzeb** [reply]
It is in Laravel 11. Taylor presented it at Laracon US. https://laravel.com/docs/11.x/container#contextual-attributes
